### PR TITLE
fix(router): fix #49457 outlet activating with old info

### LINKED ISSUE #49457: Router outlet inside ngIf getting activated with old route
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

No

### Description

When child router-outlet is created after route is changed to no longer include this child, then the outlet remains incorrectly activated.

In the example below, parent component initialization takes 1000ms. If router navigates to 'Child' and then back to 'Parent' in lower time (500ms), then router-outlet with child gets activated at 1000ms even though the route was already changed. Since the outlet is activated, navigating back to 'Child' results in exception.

If I change the time from 500ms to more than 1000ms, then child outlet is correctly deactivated.

```
  breakApp() {
    this.router.navigate([`Parent/Child`], { relativeTo: this.activatedRoute });
    setTimeout(() => {
      this.router.navigate(['Parent'], { relativeTo: this.activatedRoute });
    }, 500);
  }

///////////////////////////////////////////////////

@Component({
  template: `<p>parent</p>
            <div *ngIf="initialized">
              <router-outlet></router-outlet>
            </div>`
})
export class ParentComponent {
  initialized = false;

  ngOnInit(): void {
    setTimeout(() => {
      this.initialized = true;
    }, 1000);
  }
}

///////////////////////////////////////////////////

@Component({
  template: '<p>child</p>',
})
export class ChildComponent { }

///////////////////////////////////////////////////

const routes: Routes = [
  {
    path: '',
    children: [
      {
        path: 'Parent',
        component: ParentComponent,
        children: [
          {
            path: 'Child',
            component: ChildComponent,
          },
        ],
      },
    ],
  },
];
@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule],
})
export class AppRoutingModule { }
```

The issue can be seen in provided stackblitz demo. Clicking Break button results in child outlet remaining in activated state. In such state it is not possible to switch between children. After navigating back to root page Parent component gets destroyed and problem is resolved.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-dedznl

### Please provide the exception or error you saw

```true
Error: NG04013: Cannot activate an already activated outlet
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 15.1.6        
Node: 16.18.1
Package Manager: npm 8.19.2
OS: win32 x64

Angular: 15.2.3
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1502.4
@angular-devkit/build-angular   15.2.4
@angular-devkit/core            15.2.4
@angular-devkit/schematics      15.1.6
@angular/cli                    15.1.6
@schematics/angular             15.1.6
rxjs                            7.8.0
typescript                      4.9.5
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by jancabadaj**
Avoid activating outlet with old info if route was changed before outlet activated

Fixes #49457

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
@jancabadaj Thnx for the PR - in order to merge it we would definitively need a unit test that illustrates the failing scenario.

**Comment by jancabadaj** [comment]
Added test case. Test is failing without fix and passing with fix.

File: packages/router/src/operators/activate_routes.ts [hunk 1]
```diff
@@ -131,8 +131,12 @@ export class ActivateRoutes {
...
       context.children.onOutletDeactivated();
+    }
+
+    if (context) {
       // Clear the information about the attached component on the context but keep the reference to
-      // the outlet.
+      // the outlet. Clear even if outlet was not yet activated to avoid activating later with old
+      // info
       context.attachRef = null;
       context.route = null;
```
Comment (atscott) [inline packages/router/src/operators/activate_routes.ts:141]: nit: Can we change this to not check for `context` twice to keep the destroy logic together? Either with an early return or just another `if` inside the `if(context)`

```
    if (!context) {
      return;
    }

    if (context.outlet) {
      // Destroy the component
      context.outlet.deactivate();
      // Destroy the contexts for all the outlets that were in the component
      context.children.onOutletDeactivated();
    }
    // Clear the information about the attached component on the context but keep the reference to
    // the outlet.
    context.attachRef = null;
    context.resolver = null;
    context.route = null;
```

or

```
    if (context) {
      if (context.outlet) {
        // Destroy the component
        context.outlet.deactivate();
        // Destroy the contexts for all the outlets that were in the component
        context.children.onOutletDeactivated();
      }
      // Clear the information about the attached component on the context but keep the reference to
      // the outlet.
      context.attachRef = null;
      context.resolver = null;
      context.route = null;
    }
```

**Review by atscott — APPROVED** [review]

**Review by atscott — APPROVED** [review]

**Review by atscott — APPROVED** [review]

**Comment by atscott** [comment]
This PR was merged into the repository by commit d3018c0ee71eab2ab35aff20d95e9fa882944d14.

**Pull request closed without merging**